### DISCUSSION TITLE
Single instance with remote/server API for opening new files

### DISCUSSION THREAD

**Discussion by michaeljones** [helix-editor/helix#5107] [Ideas] [answered]
I imagine there is a ticket for this already but I'm not sure what terms to search for. I've tried 'server' and 'remote' which are the terms used in vim, I think.

Is it possible to open an instance of helix and then, from the command line, open further files in that instance rather than opening new instances? I like working and search for stuff on the command line and it would be ace to be able to just forward those files to helix to open rather than having to go back to helix and research for the relevant file.

**Comment by Ordoviz** [comment, marked as answer]
#312

**Reply by michaeljones** [reply]
Thanks, that seems to be a much large scope suggestion than what I had in mind. It was more that you could have Helix listening on a socket and be able to send a message to that socket saying "please open /path/to/file" and it would do the same thing as running `:open /path/to/file` in that editor instance itself. 

That might not be ideal for other reasons but it isn't a whole multiple-client + server architecture.

**Reply by dead10ck** [reply]
It actually is. Helix itself is the first client, and any number of others could use the socket.

**Comment by michaeljones** [comment]
I've added a hacky commit on top of the 22.12 tag which opens a unix socket as `/tmp/helix-socket` and listens for messages of the form `open <filename>` and triggers the `:open <filename>` command and does a redraw afterwards. It seems to work: https://github.com/michaeljones/helix/pull/1

I use it with a script like this:

```bash
#!/bin/bash

files=$*

# Detect if helix is already running - if so use socat to send "open <filename>" to the socket 
# otherwise make sure the socket file is deleted so that it is clear to open a new one
if ps -ef | grep -- "--title helix" | grep -v grep > /dev/null; then
  first=$1
  fullpath=$(realpath $first)
  echo "open ${fullpath}" | socat - UNIX-CONNECT:/tmp/helix-socket
  exit 0
else
  rm -f /tmp/helix-socket
fi

# Open helix in some way (this opens it in an instance of alacritty)
PYTHONPATH='' alacritty --title helix --option dimensions.columns=180 --option dimensions.lines=50 --command hx $files &
```
I hope this is of some use to people in my position. I have no expectations of it being merged. It is a hacky mess. It also only compiles on unix.

**Reply by michaeljones** [reply]
This needs more work if you use helix as your $EDITOR for things like `git commit`. I've just installed the plain 22.12 release as `hx-22.12` and set my $EDITOR to that to avoid issues until I have the energy to look at it more closely.

**Reply by michaeljones** [reply]
Doesn't currently ensure that the paths are absolute which they should be as they won't always be relative to the cwd of the editor.

**Reply by michaeljones** [reply]
Updated for the 23.03 release here: https://github.com/michaeljones/helix/pull/2

**Reply by michaeljones** [reply]
Updated for the 23.05 release here: https://github.com/michaeljones/helix/pull/3

**Reply by michaeljones** [reply]
Updated for the 23.10 release here: https://github.com/michaeljones/helix/pull/4

**Reply by michaeljones** [reply]
Updated for the 24.03 release here: https://github.com/michaeljones/helix/pull/5

**Reply by michaeljones** [reply]
Updated for the 24.07 release here: https://github.com/michaeljones/helix/pull/6

**Comment by kirawi** [comment]
I'm working on #401, and part of my proposal for local client-server is to have a single master client containing all the respective data. On Unix-based systems, this would be a daemon, but Windows doesn't have a direct equivalent to them. Instead, when the master client exits, one the child clients loads the persistent state and becomes the master.
